grant-project/zcash-grant-system
# Flexible Matching/Bounties + KYC

Closes #251 and closes #276.

### What this does
- adds `proposal.rfp_opt_in` - important during proposal creation
- adds `proposal.contribution_bounty`  - this is now independent of RFP for proposals
- if `rfp_opt_in` is set to true during creation, then `contribution_bounty` and `contribution_matching` get set from linked RFP (setting `rfp_opt_in` to false sets bounty and matching to zero)
- allows admin to set bounty so long as the proposal is still in funding state
- makes sure that RFP bounty is set to `None` if `null`, `''` or `'0'` is passed during update/create

### Steps to test
- upgrade your DB
- create an RFP with both matching and bounty
- create a proposal from that RFP
- you should be required to either opt-in or out of the RFP matching and bounty before submitting
- choose to opt-in, submit, approve and publish the proposal
- ensure the proposal renders with the ribbon in the list, and with both ribbons in the detail campaign block
- admin: navigate to the proposal and "Set bounty", make sure it is for less than the target
- ensure that the altered bounty renders instead of the RFP bounty
- admin: now "Set bounty" such that the target is exceeded
- admin: "matching" and "Set bounty" should now be disabled as the proposal has been fully funded

## Screens

### Create: RFP KYC Opt-in (first step)
![screen shot 2019-02-28 at 9 00 32 pm](https://user-images.githubusercontent.com/11430954/53613548-ecc7d680-3b9b-11e9-8a14-d1d73c3502b0.png)

### Create: RFP KYC Required (review step)
![screen shot 2019-02-28 at 9 01 49 pm](https://user-images.githubusercontent.com/11430954/53613601-24368300-3b9c-11e9-8c6d-a3ea26d4b2f1.png)

### Create: RFP KYC Accepted (review step)
![screen shot 2019-02-28 at 9 02 16 pm](https://user-images.githubusercontent.com/11430954/53613657-534cf480-3b9c-11e9-8f6e-4e64c4ed8810.png)

### Admin: Proposal Detail > Set bounty
![screen shot 2019-02-28 at 9 04 43 pm](https://user-images.githubusercontent.com/11430954/53613714-8f805500-3b9c-11e9-83b5-87f60a189f02.png)
